mirrord/cli/src/container.rs

```diff
@@ -0,0 +1,288 @@
+
+            ContainerError::UnseccesfulCommandOutput(format_command(command), message)
+        })
+}
+
+/// Create a temfile with a json serialized [`LayerConfig`] to be loaded by container and external
+/// proxy
+#[tracing::instrument(level = Level::TRACE, ret)]
+fn create_composed_config(config: &LayerConfig) -> Result<NamedTempFile, ContainerError> {
+    let mut composed_config_file = tempfile::Builder::new()
+        .suffix(".json")
+        .tempfile()
+        .map_err(ContainerError::ConfigWrite)?;
+    composed_config_file
+        .write_all(&serde_json::to_vec(config).map_err(ContainerError::ConfigSerialization)?)
+        .map_err(ContainerError::ConfigWrite)?;
+
+    Ok(composed_config_file)
+}
+
+/// Create a tempfile and write to it a self-signed certificate with the provided subject alt names
+#[tracing::instrument(level = Level::TRACE, ret)]
+fn create_self_signed_certificate(
+    subject_alt_names: Vec<String>,
+) -> Result<(NamedTempFile, NamedTempFile), ContainerError> {
+    let geerated = rcgen::generate_simple_self_signed(subject_alt_names)
+        .map_err(ContainerError::SelfSignedCertificate)?;
+
+    let mut certificate =
+        tempfile::NamedTempFile::new().map_err(ContainerError::WriteSelfSignedCertificate)?;
+    certificate
+        .write_all(geerated.cert.pem().as_bytes())
+        .map_err(ContainerError::WriteSelfSignedCertificate)?;
+
+    let mut private_key =
+        tempfile::NamedTempFile::new().map_err(ContainerError::WriteSelfSignedCertificate)?;
+    private_key
+        .write_all(geerated.key_pair.serialize_pem().as_bytes())
+        .map_err(ContainerError::WriteSelfSignedCertificate)?;
+
+    Ok((certificate, private_key))
+}
+
+/// Create a "sidecar" container that is running `mirrord intproy` that connects to `mirrord
+/// extproxy` running on user machine to be used by execution container (via mounting on same
+/// network)
+#[tracing::instrument(level = Level::TRACE, ret)]
+async fn create_sidecar_intproxy(
+    config: &LayerConfig,
+    base_command: &RuntimeCommandBuilder,
+    connection_info: Vec<(&str, &str)>,
+) -> Result<(String, SocketAddr), ContainerError> {
+    let mut sidecar_command = base_command.clone();
+
+    sidecar_command.add_env("MIRRORD_INTPROXY_DETACH_IO", "false");
+    sidecar_command.add_envs(connection_info);
+
+    let sidecar_container_command = ContainerCommand::run([
+        "--rm",
+        "-d",
+        &config.container.cli_image,
+        "mirrord",
+        "intproxy",
+    ]);
+
+    let (runtime_binary, sidecar_args) = sidecar_command
+        .with_command(sidecar_container_command)
+        .into_execvp_args();
+
+    // We skip first index of sidecar_args because `RuntimeCommandBuilder::into_execvp_args` adds
+    // the binary as first arg for `execvp`
+    #[allow(clippy::indexing_slicing)]
+    let sidecar_container_id =
+        exec_and_get_first_line(Command::new(&runtime_binary).args(&sidecar_args[1..])).await?;
```

Comment:
Hmmm, this probably won't ever happen, but we are denying it everywhere ... I'm not sure it's worth the hassle changing this to an error and not to just panic with indexing slicing.

Do you think it's worth changing this to slice of iterator skipping the 1st element? I'll leave this up to you.

Reply:
Well it should never because the sidecar_args are generated from `std::iter::once(runtime_binary).chain(...)` but I can split the function to have one for exec and one that returns without the added arg at 0